triton/tut/parallel-shared.rst

```diff
@@ -0,0 +1,257 @@
+for the same resources.
+
+Using threads and processes at the same time can also result in
+:ref:`double-booking<threads-vs-processes>`.
+
+
+Using SLURM_CPUS_PER_TASK effectively
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The environment variable ``$SLURM_CPUS_PER_TASK`` can be utilized in multiple
+ways in your scripts. Below are few examples:
+
+- Setting a number of workers when ``$SLURM_CPUS_PER_TASK``  is not set:
+
+  ``$SLURM_CPUS_PER_TASK`` is only set when ``--cpus-per-task`` has 
+  been specified. If you want to run the same code in your own machine
+  and in the cluster it might be useful to set a variable like
+  ``export NCORES=${SLURM_CPUS_PER_TASK:-4}`` and use that in your scripts.
+
+  Here ``$NCORES`` is set to the number specified by ``$SLURM_CPUS_PER_TASK``
+  if it has been set. Otherwise, it will be set to 4.
+
+- In Python you can use the following for obtaining the environment
+  variable:
+
+  .. code-block:: python
+
+      import os
+
+      ncpus=int(os.environ.get("SLURM_CPUS_PER_TASK", 1))
+
+  For more information on parallelisation in Python see our
+  :doc:`Python documentation </triton/apps/python>`.
+
+- In R you can use the following for obtaining the environment
+  variable:
+
+  .. code-block:: r
+
+      ncpus <- as.integer(Sys.getenv("SLURM_CPUS_PER_TASK", unset=1))
+
+  For more information on parallelisation in R see our
+  :doc:`R documentation </triton/apps/r>`.
+
+- In Matlab you can use the following for obtaining the environment
+  variable:
+
+  .. code-block:: matlab
+
+      ncpus=str2num(getenv("SLURM_CPUS_PER_TASK"))
+
+  For more information on parallelisation in Matlab see our
+  :doc:`Matlab documentation </triton/apps/matlab>`.
+
+
+Lack of parallelisation when using srun
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Since Slurm version 22.05 job steps run with ``srun`` will not
+automatically inherit the ``--cpus-per-task``-value that is requested
+by ``sbatch``. This was done to make it easier to start multiple job
+steps with different CPU allocations within one job.
+
+If you want to give all CPUs to ``srun`` you can either call ``srun``
+in the script with ``srun --cpus-per-task=$SLURM_CPUS_PER_TASK`` or
+set:
+
+.. code-block:: sh
+
+     export SRUN_CPUS_PER_TASK=$SLURM_CPUS_PER_TASK
+
+
+For more information see documentation pages for
+`srun <https://slurm.schedmd.com/srun.html>`_ and
+`sbatch <https://slurm.schedmd.com/sbatch.html>`_.
+
+
+Asking for multiple tasks when code does not use MPI
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+
+Normally you should **not** use ``--ntasks=n`` when you want to
+run shared memory codes. The number of tasks is only relevant to MPI codes
+and by specifying it you might launch multiple copies of your program
+that all compete on the reserved CPUs.
+
+Only hybrid parallelization codes should have both ``--ntasks=n`` and
+``--cpus-per-task=C`` set to be greater than one.
+
+.. _effective-cpus-per-task:
+
+What's next?
+------------
+
+
```

Comment:
This is very minor, but: I usualy try to have three lines before the heading/ref marker and one after for top-level sections like this.  Lower level sections, anything goes depending on what feels right.

Reply:
Good point. Fixed in the latest commit. Also the anchor in the example was in the wrong place so I fixed that as well.